File: srcJava/com/rti/perftest/harness/AnnouncementListener.java
/*
 * (c) 2005-2017  Copyright, Real-Time Innovations, Inc. All rights reserved.
 * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
 */

package com.rti.perftest.harness;

import java.util.ArrayList;
import com.rti.perftest.IMessagingCB;
import com.rti.perftest.TestMessage;

//===========================================================================

/**
 * Data listener for the Announcement
 *
 * Receives an announcement message from a Subscriber once
 * the subscriber has discovered every Publisher.
 */
/*package*/ final class AnnouncementListener implements IMessagingCB {
    // -----------------------------------------------------------------------
    // Public Fields
    // -----------------------------------------------------------------------
    public ArrayList<Integer> subscriber_list;

    // -----------------------------------------------------------------------
    // Public Methods
    // -----------------------------------------------------------------------

    // --- Constructors: -----------------------------------------------------

    public AnnouncementListener() {
        subscriber_list = new ArrayList<Integer>();
    }

    // --- From IMessagingCB: ------------------------------------------------

    public void processMessage(TestMessage message) {
        /*
         * The subscriber_list vector contains the list of discovered subscribers.
         *
         * - If the message.size is INITIALIZE or LENGTH_CHANGED and the
         *   subscriber is not in the list, it will be added.
         * - If the message.size is FINISHED_SIZE and the
         *   subscriber is in the list, it will be removed.
         *
         * The publisher access to this list to verify:
         * - If all the subscribers are discovered or notified about the length
         *   being changed.
         * - If all the subscribers are notified that the test has finished.
         */
        if ((message.size == PerfTest.INITIALIZE_SIZE
                || message.size == PerfTest.LENGTH_CHANGED_SIZE)
                && !subscriber_list.contains(message.entity_id)) {
            subscriber_list.add(message.entity_id);
        } else if (message.size == PerfTest.FINISHED_SIZE) {
            subscriber_list.remove(new Integer(message.entity_id));
        }
    }

}

// ===========================================================================

File: srcCs/Harness/AnnouncementListener.cs
/*
 * (c) 2005-2021 Copyright, Real-Time Innovations, Inc. All rights reserved.
 * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
 */

using System.Collections.Generic;

namespace PerformanceTest
{
    public class AnnouncementListener : IMessagingCallback
    {
        public int announcedSubscriberReplies;
        private readonly List<int> finishedSubscribers = new List<int>();

        public void ProcessMessage(TestMessage message)
        {
            /*
             * If the entity_id is not in the list of subscribers
             * that finished the test, add it.
             *
             * Independently, decrease announced_subscriber_replies if a known
             * writer responds to a message using this channel. We use
             * this as a way to check that all the readers have received
             * a message written by the Throughput writer.
             */
            if (!finishedSubscribers.Contains(message.entityId))
            {
                finishedSubscribers.Add(message.entityId);
                announcedSubscriberReplies++;
            }
            else
            {
                announcedSubscriberReplies--;
            }
        }
    }
} // PerformanceTest Namespace
